### PR TITLE
support for webhdfs file system

### PR DESCRIPTION
**Pull request by harsham4026**

### CODE REVIEW

**Comment by harsham4026** [comment]
@MSeal developed the basic things that we do with HDFS. Please have a look and let me know for the changes and addition of features :)

**Comment by harsham4026** [comment]
@MSeal did you get a chance to review this PR. Looking forward to listening to your suggestions or changes :)

**Comment by harsham4026** [comment]
@betatim sure. Preparing the test cases. Soon will commit them.

File: papermill/hdfs.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+# python 2 compatibility imports
+
+try:
+    FileNotFoundError
+except NameError:
+    FileNotFoundError = IOError
+
+
+class HDFS(object):
```
Comment (MSeal) [inline papermill/hdfs.py]: We probably don't need to make all the HDFS operators prefixed with `_`. I'd say most of these are fine to expose without the private prefix.

File: papermill/hdfs.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    def _is_empty(self, path):
+        """:param path:str path which needs to be checked if it's a directory
+           :returns true if the given path is empty else false"""
+        try:
+            file_status = self.__webhdfs.get_file_dir_status(path)
+            if len(file_status) > 0:
+                return False
+            else:
+                return True
+        except Exception as e:
```
Comment (MSeal) [inline papermill/hdfs.py]: This is redundant, removing it has the same behavior

File: papermill/hdfs.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            for f in hdfs_files['FileStatuses']['FileStatus']:
+                yield path + f['pathSuffix']
+        else:
+            raise FileNotFound("directory {directory} doesn't exist".format(directory=path))
+
+    def _read(self, path, size=-1, offset=0, binary=False):
+        """:param path:str
+           :returns reads contents of the given file path and returns"""
+        if self._exists(path) and not self._is_file(path):
+            raise IOError("file {path} doesn't exist".format(path=path))
```
Comment (MSeal) [inline papermill/hdfs.py]: Let's raise FileNotFoundError for this

File: papermill/hdfs.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        read = self.__webhdfs.read_file(path, offset=offset, length='null' if size < 0 else size)
```
Comment (MSeal) [inline papermill/hdfs.py]: Note sure `length='null'` is right?

File: papermill/hdfs.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        if self._exists(path) and not self._is_file(path):
+            hdfs_files = self.__webhdfs.list_dir(path)
+            if not path.endswith('/'):
+                path += '/'
+            for f in hdfs_files['FileStatuses']['FileStatus']:
+                yield path + f['pathSuffix']
+        else:
+            raise FileNotFound("directory {directory} doesn't exist".format(directory=path))
+
+    def _read(self, path, size=-1, offset=0, binary=False):
```
Comment (MSeal) [inline papermill/hdfs.py]: Better to use `size=None` for python interfaces

File: papermill/hdfs.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        """:param path:str
+           :returns reads contents of the given file path and returns"""
+        if self._exists(path) and not self._is_file(path):
+            raise IOError("file {path} doesn't exist".format(path=path))
+        read = self.__webhdfs.read_file(path, offset=offset, length='null' if size < 0 else size)
+        if not binary:
+            read = read.decode('utf-8')
+        return read
+
+    def _write(self, path, s=""):
```
Comment (MSeal) [inline papermill/hdfs.py]: `s` doesn't need `=""`, also let's name this better as `contents`

File: papermill/hdfs.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        """:param path:str
+           :returns writes the given content to file. By default it'll overwrite the existing file's
+        contents. Can be disabled."""
+        if not self._exists(path) and self._exists(self._path_basename(path)):
+            return self.__webhdfs.create_file(path, s, overwrite=True)
+        else:
+            raise FileNotFound("parent directory {} doesn't exist".format(self._path_basename(path)))
+
+    def _mkdir(self, path, recursive, exist_ok=False):
```
Comment (MSeal) [inline papermill/hdfs.py]: You don't actually use `recursive` except to raise exceptions? I'd drop `exist_ok` and just assume it's True for execution purposes

File: papermill/hdfs.py (hunk 1 above)
Comment (harsham4026) [reply papermill/hdfs.py]: removed the private preix

File: papermill/hdfs.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+"""Utilities for working with HDFS."""
+import logging
+
+from pywebhdfs.errors import FileNotFound
+from pywebhdfs.webhdfs import PyWebHdfsClient
```
Comment (MSeal) [inline papermill/hdfs.py]: Before you get too far fixing things, I do want to point out that we should probably use pyarrow for `hdfs://` paths and pywebhdfs for `webhdfs://` paths. Not all hadoop clusters will have webhdfs installed/enabled so we'll want the register to differentiate the two.

Comment (harsham4026) [reply papermill/hdfs.py]: okay. Will make changes for pyarrow :)

Comment (MSeal) [reply papermill/hdfs.py]: If you want to leave this branch for `webhdfs` that's fine too, just wanted to let you know there was a difference between the two.

Comment (harsham4026) [reply papermill/hdfs.py]: sure

Comment (harsham4026) [reply papermill/hdfs.py]: Added the code for supporting hdfs. PR #224 addresses this.

**Comment by harsham4026** [comment]
@MSeal are we planning to webhdfs support in near future? If we are planning to add I'll try to add the test cases for the code of pywebhdfs committed.

File: papermill/hdfs.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        timeout: timeout for the underlying HTTP request
+        base_uri_pattern: format string for base URI
+        request_extra_opts: dictionary of extra options to pass
+          to the requests library (e.g., SSL, HTTP authentication, etc.)
+        >>> hdfs = PyWebHdfsClient(host='host',port='50070', user_name='hdfs')
+    """
+
+    def __init__(self, host=None, port=None, user_name=None,
+                 path_to_hosts=None, **kwargs):
+        self.__webhdfs = PyWebHdfsClient.__init__(self, host, port, user_name, **kwargs)
```
Comment (MSeal) [inline papermill/hdfs.py]: Use single `_` for private variables, double `__` is usually for hidden system attributes.

File: papermill/utils.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    def decorate(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            for i in range(num):
+                try:
+                    return func(*args, **kwargs)
+                except Exception as e:
+                    logger.debug('retrying: {}'.format(e))
+            else:
+                raise Exception  # TODO verify what to raise
```
Comment (MSeal) [inline papermill/utils.py]: I know this was just copied, but let's improve this some. This should `raise e` and the line `e = Exception("No attempts made") should be at the top of `wrapper`.

**Comment by MSeal** [comment]
> @MSeal are we planning to webhdfs support in near future? If we are planning to add I'll try to add the test cases for the code of pywebhdfs committed.

@harsham4026 I see no problem with adding this. There are times you need to use this over native hdfs calls. It will require extensive testing though, so you'll need at least one test per function you added (more for the functions with more complex logic). That'll be a little bit of work to complete, but if you get stuck on any particular test / mock pattern I'll respond with how to achieve desired testing patterns.

**Comment by harsham4026** [comment]
@MSeal I was exploring the libraries to mock the webhdfs and hdfs, can you please suggest me the library to mock hdfs for writing the test cases?

**Comment by rgbkrk** [comment]
This PR has been open long enough that I'm going to recommend trying it again as a new PR, given how much has changed since November. Feel free to mention this PR in any new PRs related to this one.

Cheers!

**Pull request closed without merging**